src/DataProtection/DataProtection/src/KeyManagement/KeyRingBasedDataProtector.cs

```diff
@@ -367,7 +367,8 @@ internal static byte[] BuildAadTemplateBytes(string[] purposes)
             var keySize = sizeof(Guid);
             int totalPurposeLen = 4 + keySize + 4;
 
-            var purposeLengthsPool = ArrayPool<int>.Shared.Rent(purposes.Length);
+            int[]? lease = null;
+            Span<int> purposeLengthsPool = purposes.Length <= 32 ? stackalloc int[purposes.Length] : (lease = ArrayPool<int>.Shared.Rent(purposes.Length)).AsSpan(0, purposes.Length);
```

Comment:
:+1: from me, interested in @sebastienros thoughts

Reply:
> I'm guessing we don't actually expect many purposes

I don't know exactly where this DP is used, in the case of antiforgery we saw a single `purpose`. But assuming this is the default in any ASP.NET it may be used for way more things. Would be good to get this data for the benchmarks, because we might just screw other scenarios. At least we could add a `[Parameter]` to the benchmark with other numbers of purposes. (1, 8, 64) for starter? But it might also be ever better for other scenarios where the allocations would be worse and this would make even more sense than just for AF.

Reply:
there shouldn't really be any allocation here - or a least, it is amortized over the pool usage; we aren't borrowing it for an extended period of time or anything, so: the only time we can really get into problems here is if the pool itself is already totally borked and drowning, in which case we'd be the victim, not the perpetrator